### ISSUE TITLE
Refactoring

### ISSUE THREAD

**Issue by alex-ht** [thu-spmi/CAT#15]
Hi, I  am trying to make a Dockerfile and refactor directory structures.

In my branch [docker](https://github.com/alex-ht/CAT/tree/docker), I have finished a [dockerfile](https://github.com/alex-ht/CAT/blob/docker/docker/Dockerfile) ~~based on kaldi's official `gpu-latest`  image. The pytorch is at 1.2.0 version, because the base image, kaldiasr/kaldi:gpu-latest, is staying at CUDA10.0. I'm not familiar with  pytorch, so I just copy the instrunction from official site [here](https://pytorch.org/get-started/previous-versions/).~~

Update:  I chose pytorch 1.5 as base image now. 

After those works, I found out that the scripts in wsj/steps and utils are not following kaldi's guidline.
Some of them were copied from old version of kaldi, Eesen, etc. I would suggest untracking typeical scripts from old Kaldi and moving remains into [scripts/ctc-crf](https://github.com/alex-ht/CAT/tree/docker/scripts/ctc-crf). Every egs has two soft links, `steps` and `utils`, to kaldi's wsj, and one soft link goes to `scripts/ctc-crf `.

Do you agree with above changes? I'm happy to make PR.

**Comment by aky15** [comment]
Thanks for your contribution to our work. The structures are more clearer now. Have you test the scripts (In particular, the dependencies between scripts) to make sure they can work?

**Comment by aky15** [comment]
And the installation guidance in readme.md need to be modified. It may be a good idea to provide optional choices for somebody who already have kaldi and pytorch installed.

**Comment by alex-ht** [comment]
It is just work in progress now. I will update aishell folder, them maybe you can follow new scripts to update wsj, and other LDC corpus? I have no membership of LDC, and no powerful machines.

**Comment by aky15** [comment]
Thank you and I'll try it.

**Comment by alex-ht** [comment]
Hi @aky15,
It the sources `src/batchnorm_src` out of the date? Can it be replaced with  something likes torch.nn.BatchNorm1D/2D?

**Comment by alex-ht** [comment]
BTW, I'm running the aishell egs. It will take me few days to get WER results.

**Comment by aky15** [comment]
BatchNorm in torch can not handle input with varying lengths, which is very common in ASR. Specifically, short utterance in the batch are padded with zero, and these zeros should not be included in the mean and variance calculation. src/batchnorm_src is meant for this.

**Comment by alex-ht** [comment]
Bother you again.

I got initial result on aishell finally. I have made two changes to make it work on my pc.
1) Batch size was reduced to 100, because I have only one GPU with 12GB ram.
2) The "batchnorm_src" was replaced with built-in BatchNorm2D. I cannot find setup script for pytorch 1.x cpp extension likes ctc-crf/setup_1_0.py.

I kill the training process after epoch 17. Here is progress report
```
$ grep mean_cv_loss  my.6.log     
mean_cv_loss: -15.59308932104776
mean_cv_loss: -16.70415916442871
mean_cv_loss: -17.101018417713252
mean_cv_loss: -17.341455113610557
mean_cv_loss: -17.45301233335983
mean_cv_loss: -17.57776860969011
mean_cv_loss: -17.65706866064737
mean_cv_loss: -17.663358138328373
mean_cv_loss: -17.653572401889534
mean_cv_loss: -17.716543197631836
mean_cv_loss: -17.751277187258697
mean_cv_loss: -17.78889011116915
mean_cv_loss: -17.68128953090934
mean_cv_loss: -17.738837211076603
mean_cv_loss: -17.70960905385572
mean_cv_loss: -18.05926601498626
mean_cv_loss: -18.084275427529978
```
And the CER result
```
$ grep WER exp/decode_test/lattice/cer_* | utils/best_wer.sh 
%WER 6.03 [ 18937 / 314295, 481 ins, 844 del, 17612 sub ] exp/decode_test/lattice/cer_9
```
It looks really good!

Do you have the batchnorm_src setup script for python1.x? I am not familiar with pytorch.

**Comment by aky15** [comment]
Hi, Alex, happy to see that you get really good results.
We will implement batchnorm_src setup for python1.x later. A possible alternative is layer norm along the feature dimension, which will not be affected by the length of the utterance.

**Comment by alex-ht** [comment]
I know many tricks to avoid the zero padding issue. I want to share this work to researchers in Taiwan, so I wish it can reproduce your research works with meaningful math as more than another deep learning tool.

**Comment by ZhijianOu** [comment]
Hi Alex,  Nice to see you got the really nice result on AISHELL.
What do you mean by saying "reproduce your research works with meaningful math"?

**Comment by alex-ht** [comment]
The batchnorm is wrong in my experiment. As aky15 says, it counts padding zeros while estimating mean and variance.

**Comment by alex-ht** [comment]
To get a reasonable result is important than others, and maybe researchers will apply or extend your work. At least my colleague have great interest in it😄

**Comment by alex-ht** [comment]
沒想到教授會親自上線，我就不用破英文荼毒大家了。
我的同事對您的研究很感興趣，但建制實驗環境遇到不少問題，所以我打算把這份程式整理好並且分享給更多在台灣的學者。現在有很多研究會忽略細節，就像zero padding的問題，大部分的人會選擇先根據長度排序，再進行mini-batch訓練。依aky15的說明，似乎是從根本上解決這問題，我覺得很值得保留下來。所以才希望aky15能提供pytorch 1.x版的安裝方式。
日後如果台灣有舉辦活動，例如競賽，就可以用您的研究作為一個baseline範例，讓更多人使用。

**Comment by alex-ht** [comment]
@aky15 我會嘗試這個方法
https://gist.github.com/amiasato/902fc14afa37a7537386f7b0c5537741#file-maskedbatchnorm1d-py

**Comment by ZhijianOu** [comment]
@alex-ht   I see. Thanks.
@aky15  Before implementing batchnorm_src setup for python1.x, you may check if the URL mentioned by alex-ht is already a correct solution.

**Comment by alex-ht** [comment]
Hi,替換後結果是CER6.10%，一樣第17 epoch後停止，相對變化只有1.1%，我認為幾乎是在誤差範圍。

另外，我發現預設情況是採用6層BLSTM的設定，並沒有受`batchnorm_src`的影響🤣。
我會在`train.py`加個參數選項選擇不同DNN架構，然後繼續測試看看。

**Comment by alex-ht** [comment]
Hi @ZhijianOu @aky15 ，
請問@aky15有確定maskedbatchnorm1d-py是否正確？我自己看是覺得無誤，只有最後一行不確定`return inp`和`return inp, lengths`是否有影響。
另外我一直無法在pytorch0.4.1的版本正確編譯batchnorm_src的程式，無法做出1.5, 0.4兩個版本供人選擇。

因為台灣這即將舉行[台灣閩南語的ASR競賽](https://sites.google.com/speech.ntut.edu.tw/fsw/home/challenge-2020)，所以來詢問一下狀況。為了趕上時間，我打算先用pytorch1.5的環境分享給台灣的ASR社群。如果教授不介意，我會用自己的fork上先分享一次。

**Comment by ZhijianOu** [comment]
@alex-ht  Thanks for the message.
I think, @aky15  can reply the questions regarding
- implementing batchnorm_src setup for python1.x
- check if maskedbatchnorm1d-py is a correct solution

I suggest you make a PR, named refactor. You may mainly do well with aishell, I will let more others to update other egs, and then do merge. Then we can work with the refactor in the future.

When is your expected time to release for use in the challenge?

**Comment by alex-ht** [comment]
Ok, 我把librispeech和台灣的formosa語料的script整理好後會先發PR

從目前的時間表來看，九月前完成會比較好
> 2020/07/01 --- Training Data Release
> 2020/09/01 --- Pilot-Test (dry-run only) Data Release
> 2020/09/15 --- Pilot-Test (dry-run only) Result Submission
> 2020/09/30 --- Pilot-Test (dry-run only) Performance Notification

**Comment by aky15** [comment]
@alex-ht Hi,Alex. In our implementation, batchnorm_src is built by default with torch 0.4* **complied from source**.  It's a bit more complicated and we met some problems to add support for its torch1.* version. We found in some datasets that layernorm yields comparable (even better) results and is more natural to solve the zero padding problem since other samples in the batch do not involved  in the mean/var computation. We have been testing the performance of  layernorm throughly and recommend to use it in our neural network model.

**Comment by alex-ht** [comment]
> layernorm

**Comment by alex-ht** [comment]
Got it with thanks, @aky15 !

**Comment by alex-ht** [comment]
另外，我无法在缩减batch size情况下让librispeech的训练收敛。大约在两三天我会发PR，再请你们试试看aishell和librispeech的script。
